components/layout_2020/display_list/mod.rs

```diff
@@ -236,6 +242,139 @@ impl DisplayListBuilder<'_> {
             self.display_list.compositor_info.epoch.as_u16(),
         ))
     }
+
+    /// Draw highlights around the node that is currently hovered in the devtools
+    fn paint_dom_inspector_highlight(
+        &mut self,
+        highlighted_dom_node: OpaqueNode,
+        fragment_tree: &FragmentTree,
+    ) {
+        const CONTENT_BOX_HIGHLIGHT_COLOR: webrender_api::ColorF = webrender_api::ColorF {
+            r: 0.23,
+            g: 0.7,
+            b: 0.87,
+            a: 0.5,
+        };
+
+        const PADDING_BOX_HIGHLIGHT_COLOR: webrender_api::ColorF = webrender_api::ColorF {
+            r: 0.49,
+            g: 0.3,
+            b: 0.7,
+            a: 0.5,
+        };
+
+        const BORDER_BOX_HIGHLIGHT_COLOR: webrender_api::ColorF = webrender_api::ColorF {
+            r: 0.2,
+            g: 0.2,
+            b: 0.2,
+            a: 0.5,
+        };
+
+        const MARGIN_BOX_HIGHLIGHT_COLOR: webrender_api::ColorF = webrender_api::ColorF {
+            r: 1.,
+            g: 0.93,
+            b: 0.,
+            a: 0.5,
+        };
+
+        let mut content_box = euclid::Rect::default();
+        let mut maybe_box_fragment = None;
+        let tag_to_find = Tag::new(highlighted_dom_node);
+        fragment_tree.find(|fragment, _, containing_block| {
+            if fragment.tag() != Some(tag_to_find) {
+                return None::<()>;
+            }
+
+            let fragment_relative_rect = match fragment {
+                Fragment::Box(fragment) | Fragment::Float(fragment) => {
+                    maybe_box_fragment = Some((fragment.clone(), *containing_block));
+
+                    fragment.borrow().content_rect
+                },
+                Fragment::Positioning(fragment) => fragment.borrow().rect,
+                Fragment::Text(fragment) => fragment.borrow().rect,
+                Fragment::Image(image_fragment) => image_fragment.borrow().rect,
+                Fragment::AbsoluteOrFixedPositioned(fragment) => {
+                    fragment.borrow().static_position_rect
+                },
+                Fragment::IFrame(iframe_fragment) => iframe_fragment.borrow().rect,
+            };
+
+            let rect = fragment_relative_rect.translate(containing_block.origin.to_vector());
+            content_box = content_box.union(&rect);
+
+            None::<()>
+        });
+        let content_box = content_box.to_webrender();
+
+        // Highlight content box
+        let properties = wr::CommonItemProperties {
+            clip_rect: content_box,
+            spatial_id: self.current_scroll_node_id.spatial_id,
+            clip_chain_id: self.current_clip_chain_id,
```

Comment:
Hrm. This is not going to work. You need to use the `spatial_id` and `clip_chain_id` that is associated with the fragment. You can observe the issue by doing a translation on an element and observing the highlight. I think you are going to have to hook into the per-Fragment display list generation routine. Each time we add a Fragment to the display list compare its tag to the highlighted tag. This might be a little annoying, but you will be able to avoid the call to `find`.